### ISSUE TITLE
Jaccard metric above 1 and black square as prediction?

### ISSUE THREAD

**Issue by EtagiBI** [qubvel/segmentation_models#319]
Hello,

I'm trying to implement jaccard loss function and jaccard metric from segmentation_models for my Unet model. 

```
# Helper functions
def _gather_channels(x, indexes):
    """Slice tensor along channels axis by given indexes"""
    if tf.keras.backend.image_data_format() == 'channels_last':
        x = tf.keras.backend.permute_dimensions(x, (3, 0, 1, 2))
        x = tf.keras.backend.gather(x, indexes)
        x = tf.keras.backend.permute_dimensions(x, (1, 2, 3, 0))
    else:
        x = tf.keras.backend.permute_dimensions(x, (1, 0, 2, 3))
        x = tf.keras.backend.gather(x, indexes)
        x = tf.keras.backend.permute_dimensions(x, (1, 0, 2, 3))
    return x


def get_reduce_axes(per_image):
    axes = [1, 2] if tf.keras.backend.image_data_format() == 'channels_last' else [2, 3]
    if not per_image:
        axes.insert(0, 0)
    return axes


def gather_channels(*xs, indexes=None):
    """Slice tensors along channels axis by given indexes"""
    if indexes is None:
        return xs
    elif isinstance(indexes, (int)):
        indexes = [indexes]
    xs = [_gather_channels(x, indexes=indexes) for x in xs]
    return xs


def round_if_needed(x, threshold):
    if threshold is not None:
        x = tf.keras.backend.greater(x, threshold)
        x = tf.keras.backend.cast(x, tf.keras.backend.floatx())
    return x


def average(x, per_image=False, class_weights=None):
    if per_image:
        x = tf.keras.backend.mean(x, axis=0)
    if class_weights is not None:
        x = x * class_weights
    return tf.keras.backend.mean(x)


# Main functions
def jaccard_metric(gt_mask, pred_mask, class_weights=1., class_indexes=None, smooth=1e-5, per_image=False, threshold=None):
    gt_mask, pred_mask = gather_channels(gt_mask, pred_mask, indexes=class_indexes)
    pred_mask = round_if_needed(pred_mask, threshold)
    axes = get_reduce_axes(per_image)

    # score calculation
    intersection = tf.keras.backend.sum(gt_mask * pred_mask, axis=axes)
    union = tf.keras.backend.sum(gt_mask + pred_mask, axis=axes) - intersection

    score = (intersection + smooth) / (union + smooth)
    score = average(score, per_image, class_weights)

    return score


def jaccard_loss(gt_mask, pred_mask, class_weights=1., class_indexes=None, smooth=1e-5, per_image=False, threshold=None):
    return 1 - jaccard_metric(gt_mask, pred_mask, class_weights=class_weights, class_indexes=class_indexes, smooth=smooth, per_image=per_image, threshold=threshold)


# Wrapper functions for meet Keras requirements
def jaccard_loss_wraper(class_weights=1., class_indexes=None, smooth=1e-5, per_image=False, threshold=None):
    def jaccard_loss_keras(gt_mask, pred_mask):
        return jaccard_loss(gt_mask, pred_mask, class_weights=class_weights, class_indexes=class_indexes, smooth=smooth, per_image=per_image, threshold=threshold)

    return jaccard_loss_keras


def jaccard_metric_wraper(class_weights=1., class_indexes=None, smooth=1e-5, per_image=False, threshold=None):
    def jaccard_metric_keras(gt_mask, pred_mask):
        return jaccard_metric(gt_mask, pred_mask, class_weights=class_weights, class_indexes=class_indexes, smooth=smooth, per_image=per_image, threshold=threshold)

    return jaccard_metric_keras


# Compilation
CLASS_WEIGHTS = [0.5, 5.]
loss_function = jaccard_loss_wraper(class_weights=CLASS_WEIGHTS, per_image=True)
metric = jaccard_metric_wraper(class_weights=CLASS_WEIGHTS, per_image=True)
model.compile(optimizer=Adam(lr=0.001), loss=loss_function, metrics=[metric])
```

I observe unusual results during training process. Metric exceeds 1, so loss becomes negative.
![Metrics](https://user-images.githubusercontent.com/20855725/78130135-2a97ce00-7432-11ea-8419-08da3fd614f0.jpg)
I feel that something's wrong here because I get black square as prediction for this model. It's either architectural flaw of mine or wrong class_weights usage.

**Comment by JordanMakesMaps** [comment]
I think it might be due to the values of your weights being greater than one, and/or also calculating the loss and accuracy per image while training. 

I have weight values exceeding one for some of my classes, but I do not have this issue (though I have seen negative loss using this repo before, I just can't remember what all the parameters were set to that made that happen)

Hope this helps:

See issues #294, #183, #177, #95

**Comment by EtagiBI** [comment]
> I think it might be due to the values of your weights being greater than one, and/or also calculating the loss and accuracy per image while training.

Thank you! Removing weights that exceed 1.0 actually helped me to overcome negative loss issue.
